include/cppkafka/utils/buffered_producer.h

```diff
@@ -661,25 +672,44 @@ bool BufferedProducer<BufferType, Allocator>::flush(std::chrono::milliseconds ti
     if (preserve_order) {
         CounterGuard<size_t> counter_guard(flushes_in_progress_);
         QueueType flush_queue; // flush from temporary queue
+        QueueType retry_flush_queue; // flush from temporary retry queue
         {
             std::lock_guard<std::mutex> lock(mutex_);
+            std::swap(retry_messages_, retry_flush_queue);
             std::swap(messages_, flush_queue);
         }
+        auto queue_flusher = [this](QueueType& queue)->bool
+        {
+            if (!queue.empty()) {
+                sync_produce(queue.front());
+                queue.pop_front();
+                return true;
+            }
+            return false;
+        };
         auto remaining = timeout;
         auto start_time = std::chrono::high_resolution_clock::now();
         do {
-            sync_produce(flush_queue.front());
-            flush_queue.pop_front();
+            if (!queue_flusher(retry_flush_queue) && !queue_flusher(flush_queue)) {
```

Comment:
I think this statement should be `!(queue_flusher(retry_flush_queue) || queue_flusher(flush_queue))` because the way it's written now both flushers will be evaluated which means you will flush alternatively from each queue. The retry queue should be fully flushed _before_ the main queue.

Reply:
The condition statements "(!queue_flusher(retry_flush_queue) && !queue_flusher(flush_queue)" and "!(queue_flusher(retry_flush_queue) || queue_flusher(flush_queue))" are equivalent. 

In the former condition, if queue_flusher(retry_flush_queue) returns true, then the negation of it will evaluate to false, hence because of the short-circuit implementation, the second part of the condition queue_flusher(flush_queue) will not be evaluated because it's clear that the entire statement will evaluate to false regardless. So queue_flusher(flush_queue) will be evaluated if and only if queue_flusher(retry_flush_queue) returns false.

However, if you think that changing the way you suggested will improve the readability, I'm ready to do that.

Reply:
Yes you're right, that makes sense. I somehow only thought of the false case.